accumulator/pollardproof.go

```diff
@@ -79,3 +75,139 @@ func (p *Pollard) IngestBatchProof(bp BatchProof) error {
+		return true, nil
+	}
+
+	// Construct a map with positions to hashes
+	proofmap, err := bp.Reconstruct(numLeaves, rows)
+	if err != nil {
+		fmt.Printf("VerifyBlockProof Reconstruct ERROR %s\n", err.Error())
+		return false, proofmap
+	}
+
+	rootPositions, rootRows := getRootsReverse(numLeaves, rows)
+
+	// partial forest is built, go through and hash everything to make sure
+	// you get the right roots
+
+	tagRow := bp.Targets
+	nextRow := []uint64{}
+	sortUint64s(tagRow) // probably don't need to sort
+
+	// TODO it's ugly that I keep treating the 0-row as a special case,
+	// and has led to a number of bugs.  It *is* special in a way, in that
+	// the bottom row is the only thing you actually prove and add/delete,
+	// but it'd be nice if it could all be treated uniformly.
+
+	if verbose {
+		fmt.Printf("tagrow len %d\n", len(tagRow))
+	}
+
+	var left, right uint64
+
+	// iterate through rows
+	for row := uint8(0); row <= rows; row++ {
+		// iterate through tagged positions in this row
+		for len(tagRow) > 0 {
+			//nextRow = make([]uint64, len(tagRow))
+			// Efficiency gains here. If there are two or more things to verify,
+			// check if the next thing to verify is the sibling of the current leaf
+			// we're on. Siblingness can be checked with bitwise XOR but since targets are
+			// sorted, we can do bitwise OR instead.
+			if len(tagRow) > 1 && tagRow[0]|1 == tagRow[1] {
+				left = tagRow[0]
+				right = tagRow[1]
+				tagRow = tagRow[2:]
+			} else { // if not only use one tagged position
+				right = tagRow[0] | 1
+				left = right ^ 1
+				tagRow = tagRow[1:]
+			}
+
+			if verbose {
+				fmt.Printf("left %d rootPoss %d\n", left, rootPositions[0])
+			}
+			// If the current node we're looking at this a root, check that
+			// it matches the one we stored
+			if left == rootPositions[0] {
+				if verbose {
+					fmt.Printf("one left in tagrow; should be root\n")
+				}
+				// Grab the received hash of this position from the map
+				// This is the one we received from our peer
+				computedRootHash, ok := proofmap[left]
+				if !ok {
+					fmt.Printf("ERR no proofmap for root at %d\n", left)
+					return false, nil
+				}
+				// Verify that this root hash matches the one we stored
+				if computedRootHash != roots[0] {
+					fmt.Printf("row %d root, pos %d expect %04x got %04x\n",
+						row, left, roots[0][:4], computedRootHash[:4])
+					return false, nil
+				}
+				// otherwise OK and pop of the root
+				roots = roots[1:]
+				rootPositions = rootPositions[1:]
+				rootRows = rootRows[1:]
+				break
+			}
+
+			// Grab the parent position of the leaf we've verified
+			parentPos := parent(left, rows)
+			if verbose {
+				fmt.Printf("%d %04x %d %04x -> %d\n",
+					left, proofmap[left], right, proofmap[right], parentPos)
+			}
+			var parhash Hash
+			n, _, _, err := p.readPos(parentPos)
+			if err != nil {
+				if err == ErrorStrings[ErrorNoPollardNode] {
+					panic(err)
+				} else {
+					panic(err)
+				}
```

Comment:
i guess you had the if statement here for testing, could remove it and just panic

Reply:
will fix